editor/find_in_files_scanner.cpp

```diff
@@ -0,0 +1,352 @@
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "find_in_files_scanner.h"
+
+const int SCRIPT_SOURCE_TAG_LEN = String("script/source = \"").length();
+
+Vector<FindInFilesScanner::ScanMatch> FindInFilesScanner::scan(String file_path, String pattern, bool match_case, bool whole_words) {
+	Ref<FileAccess> file = FileAccess::open(file_path, FileAccess::READ);
+	Vector<FindInFilesScanner::ScanMatch> matches;
+	if (file.is_null()) {
+		print_verbose(String("Cannot open file ") + file_path);
+		return matches;
+	}
+
+	String extension = file->get_path().get_extension().to_lower();
+	if (extension == "tscn") {
+		BuiltinScriptFindInFilesScanner::scan(file, file_path, pattern, match_case, whole_words, matches);
+	} else {
+		// No specific match was found, use default scanner
+		DefaultFindInFilesScanner::scan(file, file_path, file_path, pattern, match_case, whole_words, matches);
+	}
+
+	return matches;
+}
+
+void FindInFilesScanner::replace(String relative_path, String absolute_path, Vector<ScanLocation> &locations, bool match_case, bool whole_words, String search_text, String new_text) {
+	Ref<FileAccess> file = FileAccess::open(absolute_path, FileAccess::READ);
+	if (file.is_null()) {
+		print_verbose(String("Cannot open file ") + file->get_path());
+		return;
+	}
+
+	String extension = absolute_path.get_extension().to_lower();
+	if (extension == "tscn") {
+		BuiltinScriptFindInFilesScanner::replace(file, relative_path, locations, match_case, whole_words, search_text, new_text);
+	} else {
+		// No specific match was found, use default scanner
+		DefaultFindInFilesScanner::replace(file, relative_path, locations, match_case, whole_words, search_text, new_text);
+	}
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+void DefaultFindInFilesScanner::scan(Ref<FileAccess> file, String file_path, String display_text, String pattern, bool match_case, bool whole_words, Vector<FindInFilesScanner::ScanMatch> &matches, Size2i range) {
+	// Line number starts at 1.
+	for (int line_number = 1, begin = 0, end = 0; !file->eof_reached(); line_number++, begin = 0, end = 0) {
+		String line = file->get_line();
+		if (line_number < range.x) {
+			continue;
+		} else if (range.y > 0 && line_number > range.y) {
+			break;
+		}
+
+		while (find_next(line, pattern, end, match_case, whole_words, begin, end)) {
+			matches.append(FindInFilesScanner::ScanMatch{ file_path, display_text, line_number - range.x + 1, begin, end, line });
+		}
+	}
+}
+
+// Same as get_line, but preserves line ending characters.
+class ConservativeGetLine {
```

Comment:
It should probably be moved to the header with only the method definition remaining here. This ad-hociness doesn't serve any practical purpose and doesn't fit our general style. It seems Zylann, who originally implemented the feature, prefers it, judging by his other work on the engine, but since you're refactoring, it makes sense to bring everything to the same style.

If this even needs to be a dedicated class at all. It probably doesn't need to with your improved code structure. Can be just a method and a property of DefaultFindInFilesScanner.